mungegithub/mungers/publish_scripts/util.sh

```diff
@@ -19,118 +19,551 @@
-    git checkout upstream-kube/"${src_branch}" -b kube-sync
-    git reset --hard upstream-kube/"${src_branch}"
-    
-    # this command rewrites git history to *only* include $subdirectory 
-    git filter-branch -f --msg-filter 'awk 1 && echo && echo "Kubernetes-commit: ${GIT_COMMIT}"' \
-        --subdirectory-filter "${subdirectory}" HEAD
-
-    local newKubeSHA=$(git log kube-sync -1 | tail -n 1 | sed "s/Kubernetes-commit: //g")
-
-    local previousBranchSHA=$(git log --grep "Kubernetes-commit: ${previousKubeSHA}" --format='%H')
-    local commits=$(git log --no-merges --format='%H' --reverse ${previousBranchSHA}..HEAD)
-    if [ "${new_repo}" = "true" ]; then
-        commits=$(git log --no-merges --format='%H' --reverse HEAD)
-    fi
-
-    # check if any commit of commits change Godeps/Godeps.json
-    local commits_change_godeps=$(git log --format='%H' --follow Godeps/Godeps.json)
-    local cherrypicks_change_godeps="false"
-    while read commit; do
-        if [[ -z "${commit}" ]]; then
-            continue
+    local dst_branch="${3}"
+    local kubernetes_remote="${4:-https://github.com/kubernetes/kubernetes.git}"
+    local deps="${5:-""}"
+    local is_library="${6}"
+    readonly subdirectory src_branch dst_branch kubernetes_remote deps is_library
+
+    local new_branch="false"
+    if ! git rev-parse -q --verify HEAD; then
+        echo "Found repo without ${dst_branch} branch, creating initial commit."
+        git commit -m "Initial commit" --allow-empty
+        new_branch="true"
+    else
+        echo "Starting at existing ${dst_branch} commit $(git rev-parse HEAD)."
+    fi
+
+    local dst_old_head=$(git rev-parse HEAD) # will be the initial commit for new branch
+
+    # fetch upstream kube and checkout $src_branch, name it filtered-branch
+    git remote rm upstream-kube >/dev/null || true
+    git remote add upstream-kube "${kubernetes_remote}" >/dev/null
+    git fetch -q upstream-kube --no-tags
+    git branch -D filtered-branch >/dev/null || true
+    git branch -f upstream-branch upstream-kube/"${src_branch}"
+    echo "Checked out k8s.io/kubernetes commit $(git rev-parse upstream-branch)."
+    git checkout -q upstream-branch -b filtered-branch
+    git reset -q --hard upstream-branch
+
+    # filter filtered-branch (= ${src_branch}) by ${subdirectory} modifying commits
+    # and rewrite paths. Each filtered commit (which is not dropped), gets the
+    # original k8s.io/kubernetes commit hash in the commit message as "Kubernetes-commit: <hash>".
+    # Then select all new mainline commits on filtered-branch as ${f_mainline_commits}
+    # to loop through them later.
+    local f_mainline_commits=""
+    if [ "${new_branch}" = "true" ] && [ "${src_branch}" = master ]; then
+	# new master branch
+	filter-branch ${src_branch} "${subdirectory}" filtered-branch
+
+	# find commits on the main line (mostly merges, some commits due to fast-forward merges that filter-branch dropped)
```

Comment:
> some commits due to fast-forward merges that filter-branch dropped

Is this more readable: "some non-merge commits because filter-branch dropped fast-forward merges."

Reply:
improved